sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeJoinExec.scala

```diff
@@ -617,6 +619,7 @@ case class SortMergeJoinExec(
 
     s"""
        |while (findNextInnerJoinRows($leftInput, $rightInput)) {
+       |  ${leftVarDecl.mkString("\n")}
```

Comment:
can't we define them before the loop and reuse them?

Reply:
It could be. Would it be possible to let us know advantages compare to the current method?  
I think that to shorten lifetime of variables (i.e. current approach) make generated code more readable.

Reply:
Since I have no strong preference, I would like to hear opinions from others.

Reply:
I think that the advantage would be to reuse them, avoiding creating and destroying them at every loop.

Reply:
Since they are local variable, it takes almost no cost. If they are on CPU registers, there is no cost. If they are in stack frame, it is up to one instruction to increase or decrease stack frame size.

WDYT? Did you see huge overhead to create and destroy local variables?

Reply:
I am not sure about the amount of overhead introduced honestly. But I think that there might be, even if low. Therefore my suggestion was to avoid the overhead, since IMHO it is feasible and it reflects the previous situation. Honestly I don't think that the readability of the generated code is a big point, because I think that the generated code is already nearly impossible to be read as of now (without reading and knowing the code which generates it).

But maybe we can create a test and evaluate the overhead if you think this is the best option. WDYT?

Reply:
Do you think what microbench is relevant for this measurement? I would appreciate your thought.

Reply:
I was thinking about an easy benchmark like what we did for casting. WDYT?

Reply:
We would appreciate it if you would past the kernel code what you are thinking about.

Reply:
Here is my benchmark result. Is there any your result?

```
$ cat Loop.java 
public class Loop {
  private static int _i;
  private static boolean _b;
 
  public static void main(String[] args){
    Loop l = new Loop();
    long s, e;
    long c1 = 10000000000L;
    long c2 = 100000L;
    int N = 10;

    // warmup
    for (int i = 0; i < 1000000; i++) {
      l.inner(10000, 100);
      l.outer(10000, 100);
    }

    for (int n = 0; n < N; n++) {
      s = System.nanoTime();
      l.inner(c1, c2);
      e = System.nanoTime();
      System.out.println("inner(ms): " + (e-s) / 1000000);
    }

    for (int n = 0; n < N; n++) {
      s = System.nanoTime();
      l.outer(c1, c2);
      e = System.nanoTime();   
      System.out.println("outer(ms): " + (e-s) / 1000000);
    }
  }
 
  public void inner(long c1, long c2) {
    for (long i = 0; i < c1; i++) {
      int v1 = 1;
      boolean v2 = true;
      for (long j = 0; i < c2; i++) {
        _i = v1;
        _b = v2;
      }
    }
  }
 
  public void outer(long c1, long c2) {
    int v1 = 1;
    boolean v2 = true;
    for (long i = 0; i < c1; i++) {
      for (long j = 0; i < c2; i++) {
        _i = v1;
        _b = v2;
      }
    }
  }
}
$ java -version
openjdk version "1.8.0_131"
OpenJDK Runtime Environment (build 1.8.0_131-8u131-b11-2ubuntu1.16.04.3-b11)
OpenJDK 64-Bit Server VM (build 25.131-b11, mixed mode)
$ java Loop
inner(ms): 2779
inner(ms): 2779
inner(ms): 2779
inner(ms): 2779
inner(ms): 2779
inner(ms): 2779
inner(ms): 2779
inner(ms): 2779
inner(ms): 2779
inner(ms): 2779
outer(ms): 2779
outer(ms): 2779
outer(ms): 2779
outer(ms): 2779
outer(ms): 2779
outer(ms): 2779
outer(ms): 2779
outer(ms): 2779
outer(ms): 2779
outer(ms): 2779
```

Reply:
If no performance difference, we prefer to the simpler codes.

Reply:
Since these local variables are only needed inside the loop, I feel the current code is more readable. Performance is not a concern here as the overhead is very low or none.

Reply:
@kiszk thanks for the test. Then I agree this is the best option, thanks.